### ISSUE TITLE
Convolutional reverb

### ISSUE THREAD

**Issue by pseeth** [nussl/nussl#172]
**Is your feature request related to a problem? Please describe.**
It'd be nice to add the ability to convolve signals with impulse responses, possibly multichannel impulse responses.

**Describe the solution you'd like**
I think we could implement it with the following API: 

```python
import nussl

audio = nussl.AudioSignal('/path/to/audio.wav')

# load impulse from file
impulse = nussl.AudioSignal('/path/to/impulse.wav')
# or load from array gotten by some means
impulse_array = generate_random_ir()
impulse = nussl.AudioSignal(audio_data_array=impulse_array, sample_rate=audio.sample_rate)

# non destructive (default), returns a new signal
convolved = audio.convolve(impulse) # -> AudioSignal

# destructive
audio.convolve(impulse, overwrite=True) # -> inplace
```

If `impulse` is single channel, then it is applied to all channels in `audio`. If `impulse` is multi-channel, then `audio` and `impulse` must have the same number of channels. Since `impulse` is an AudioSignal object, one could do `to_mono()` if there is a channel mismatch, then apply it:

```python
# fix for channel mismatch
convolved = audio.convolve(impulse.to_mono())
```

or use `make_audio_signal_from_channel`:

```python
convolved = audio.convolve(impulse.make_audio_signal_from_channel(0))
```

Treating impulse responses as just AudioSignal possibly gives a lot of nice benefits. Users could also convolve two signals even if one of them is not an impulse response, which could sound...interesting.

**Describe alternatives you've considered**
Implementing this outside is nussl is simple by just editing the audio_data arrays. But this could be a nice feature, especially for data augmentation.

**Additional context**

**Comment by pseeth** [comment]
For actual implementation, we can just use `scipy.signal.convolve`, I think? We could do a GPU implementation via torch as well but I don't know if that's worth it.

**Comment by ethman** [comment]
As per our offline discussions of this, I think this is a good proposal. It shouldn't be that hard to implement by adding our own logic around scipy's version, which is already a dependency.

**Comment by pseeth** [comment]
Cool, I can get this going. @abugler, any thoughts?

**Comment by pseeth** [comment]
One more note, I think this could go as a function that takes two AudioSignal objects in `mixing.py` perhaps? And the `AudioSignal.convolve` can call that function with `self` and `other`. Thoughts?

**Comment by ethman** [comment]
Yeah, I wouldn't think that'd be anything but a wrapper though, right? I don't think it's inappropriate to have it in that file as well; it'd be nice to have another option for people. That being said, it is more development overhead (docs, tests, etc) to support. It's up to you.

**Comment by abugler** [comment]
> Cool, I can get this going. @abugler, any thoughts?

I think this is a cool idea! I agree that this function should take two audio signals, and the  AudioSignal hook calls it.

**Comment by ethman** [comment]
Should this have a footprint in `effects.py` somehow? I'm not sure how that would work exactly, but it does seem similar to the rest of the effects hooks.

I hesitate to fully endorse this because I don't think it should exist in `mixing.py`, `effects.py` _and_ `audio_signal.py`, but I could understand a case for one of `mixing.py` or `effects.py` honestly.

Thoughts?

**Comment by pseeth** [comment]
I think I'd like it to be in`mixing.py` with a hook in `audio_signal.py`.
